# Investigate NA values in our genotype data

In [1]:
library(CpGWAS)
library(data.table)

## Load complete SNP dataset from plink2 raw file

This raw file we're loading was produced via `plink2 --pfile libd_chr1 --export A --out libd_chr1_test`

In [2]:
file <- fread("~/data/libd_chr1_test.raw")

# Subset to columns with at least one NA
na_columns <- cbind(file$IID, file[, .SD, .SDcols = colSums(is.na(file)) > 0])

In [3]:
dim(file)

[1]   2189 326019

In [4]:
na_columns

V1     chr1:5215377:C:T_C chr1:165457376:C:A_C chr1:234308448:A:G_A
1    Br1602 2                  1.009                2.000               
2    Br1203 2                  2.000                2.000               
3    Br1214 2                  2.000                2.000               
4    Br2149 2                  2.000                1.000               
5    Br1016 2                  2.000                2.000               
6    Br1580 2                  2.000                2.000               
7    Br1646 2                  2.000                2.000               
8    Br1823 2                  2.000                1.002               
9    Br1696 2                  2.000                2.000               
10   Br1513 2                  2.000                2.000               
11   Br1930 2                  2.000                2.000               
12   Br1316 2                  1.000                1.933               
13   Br2035 2                  2.000                2.000               
14   Br2042 2                  2.000                2.000               
15   Br2129 2                  2.000                1.000               
16   Br2114 2                  2.000                2.000               
17   Br1683 2                  2.000                2.000               
18   Br1686 2                  2.000                1.001               
19   Br1436 2                  2.000                2.000               
20   Br2175 2                  2.000                1.990               
21   Br1682 2                  2.000                2.000               
22   Br1753 2                  2.000                2.000               
23   Br1761 2                  2.000                2.000               
24   Br1373 2                  1.989                2.000               
25   Br2236 2                  1.000                2.000               
26   Br2133 2                  2.000                2.000               
27   Br1980 2                  2.000                2.000               
28   Br1920 2                  1.999                2.000               
29   Br1724 2                  2.000                2.000               
30   Br2197 2                  2.000                2.000               
⋮    ⋮      ⋮                  ⋮                    ⋮                   
2160 Br847  NA                 NA                   NA                  
2161 Br890  NA                 NA                   NA                  
2162 Br1160 NA                 NA                   NA                  
2163 Br1297 NA                 NA                   NA                  
2164 Br1164 NA                 NA                   NA                  
2165 Br1285 NA                 NA                   NA                  
2166 Br845  NA                 NA                   NA                  
2167 Br1556 NA                 NA                   NA                  
2168 Br1604 NA                 NA                   NA                  
2169 Br1407 NA                 NA                   NA                  
2170 Br1227 NA                 NA                   NA                  
2171 Br1006 NA                 NA                   NA                  
2172 Br863  NA                 NA                   NA                  
2173 Br1964 NA                 NA                   NA                  
2174 Br926  NA                 NA                   NA                  
2175 Br1786 NA                 NA                   NA                  
2176 Br848  NA                 NA                   NA                  
2177 Br1342 NA                 NA                   NA                  
2178 Br1118 NA                 NA                   NA                  
2179 Br1069 NA                 NA                   NA                  
2180 Br1615 NA                 NA                   NA                  
2181 Br1348 NA                 NA                   NA                  
2182 Br1111 NA                 NA                   NA         

Do these columns (SNPs) contain NA for the same exact samples?

In [5]:
all(rowSums(is.na(na_columns[, -1, with = FALSE])) %in% c(0, ncol(na_columns) - 1))

[1] TRUE

Which indices, and which samples?

In [6]:
length(which(is.na(na_columns$`chr1:165457376:C:A_C`), arr.ind = TRUE))

[1] 531

In [7]:
2189-1659+1

[1] 531

<div class="alert alert-block alert-info">The last 531 genotypes in the SNP set have NA for these 3 SNPs.</div>

## Load `methylationInput` object containing partial SNP data loaded into R, methylation data, and various metadata

This function exists to reload data after subsetting to random samples in another session, without breaking pointers to SNP data on hard drive.

In [8]:
reinitializeMethylationInput <- function(rds_path, snp_data_path, no_cores = detectCores()) {
  if (!file.exists(rds_path)) {
    stop("RDS file does not exist: ", rds_path)
  }
  
  # Load the MethylationInput object from RDS
  loadedObject <- readRDS(rds_path)
  if (!inherits(loadedObject, "MethylationInput")) {
    stop("Loaded object is not a MethylationInput.")
  }
  
  # Reinitialize external pointers for SNP data
  pgen_path <- gsub(snp_data_path, pattern = "pvar", replacement = "pgen")
  pvar_path <- gsub(snp_data_path, pattern = "pgen", replacement = "pvar")
  psam_path <- gsub(pvar_path, pattern = "pvar", replacement = "psam")
  
  if (!file.exists(pgen_path) || !file.exists(pvar_path) || !file.exists(psam_path)) {
    stop("One or more SNP data files not found at the specified paths.")
  }
  
  loadedObject@pvar_pointer <- pgenlibr::NewPvar(pvar_path)
  loadedObject@pvar_dt <- fread(pvar_path)[, 1:3]
  loadedObject@pgen <- pgenlibr::NewPgen(pgen_path, pvar = loadedObject@pvar_pointer)
  loadedObject@psam <- fread(psam_path)
  
  # Reinitialize genotype_IDs based on intersection with methylations
  psam_in_wgbs <- loadedObject@psam[which(loadedObject@psam$`#IID` %in% rownames(loadedObject@methylations))]
  genotype_IDs <- psam_in_wgbs$`#IID`
  genotype_IDs <- intersect(rownames(loadedObject@methylations), genotype_IDs)
  loadedObject@genotype_IDs <- genotype_IDs[order(genotype_IDs)]
  
  # Ensure methylations are filtered and ordered according to the new genotype_IDs, if necessary
  loadedObject@methylations <- loadedObject@methylations[which(rownames(loadedObject@methylations) %in% genotype_IDs), ]
  
  return(loadedObject)
}

In [9]:
setwd("~/data/")

In [10]:
methInput <- reinitializeMethylationInput(rds_path = "chr1_AA_methylation_10k_samples.rds",
                                          snp_data_path = "libd_chr1.pgen",
                                          no_cores = 4)

Among samples for which we have methylation data, how many are in the the group of genotypes that has NA for the three mysterious Chr1 SNPs?

In [11]:
length(intersect(methInput@genotype_IDs, na_columns$V1[1659:2189]))

[1] 111

<div class="alert alert-block alert-info">All 111 of the genotypes in our dataset are found in the last 531 rows of the SNP set, which contain NA for all values of these three SNPs.</div>